### PR TITLE
Feat: Support macros in the MODEL statement

### PR DESCRIPTION
**Pull request by izeigerman**
This update allow users to easily parameterize their model names and other attributes by using macros and variables inside the `MODEL` statement. Thus the following example is now a valid definition:
```
MODEL (
    name @{gateway}.test_model
);
```

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1435,6 +1435,38 @@ def load_sql_based_model(
...
+        meta,
+        dialect,
+        [],
+        path=path,
+        jinja_macro_registry=jinja_macros,
+        python_env=meta_python_env,
+        default_catalog=default_catalog,
+        quote_identifiers=False,
+    )
+    rendered_meta_exprs = meta_renderer.render()
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1461]: is this expensive?

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -1888,6 +1925,20 @@ def _python_env(
                         used_macros[name] = macros[name]
                     elif name in variables:
                         used_variables.add(name)
+                elif (
+                    macro_func_or_var.__class__ is exp.Identifier and "@" in macro_func_or_var.this
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1929]: isinstance(macro_func_or_var, exp.Identifier)

File: sqlmesh/core/model/definition.py [hunk 3]
```diff
@@ -1888,6 +1925,20 @@ def _python_env(
...
+                ):
+                    var_name = None
+                    macro_template = macro_func_or_var.this
+                    at_idx = macro_template.index("@")
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1933]: what if there are two of them? are you sure you don't wanna use regex?

```
x_@{y}_@{z}

Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1933]: ah, good point. Wanted to save on regex

File: sqlmesh/core/model/definition.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1461]: Well, it's not parsing

File: sqlmesh/core/model/definition.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1929]: Keeping it consistent with the other 2 branches. Guess I should change all of them.

File: sqlmesh/core/model/definition.py (hunk 3 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py]: Turned out I could just use the pattern compiled as part of the template class construction. Fixed this

File: sqlmesh/core/model/definition.py [hunk 4]
```diff
@@ -1867,8 +1904,8 @@ def _python_env(
     expressions = ensure_list(expressions)
     for expression in expressions:
         if not isinstance(expression, d.Jinja):
-            for macro_func_or_var in expression.find_all(d.MacroFunc, d.MacroVar):
-                if macro_func_or_var.__class__ is d.MacroFunc:
+            for macro_func_or_var in expression.find_all(d.MacroFunc, d.MacroVar, exp.Identifier):
+                if isinstance(macro_func_or_var, d.MacroFunc):
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1908]: sorry i thing i may have been wrong here, the problem is that macrosql and macro def are subclasses of macrofunc

File: sqlmesh/core/model/definition.py [hunk 5]
```diff
@@ -1882,12 +1919,19 @@ def _python_env(
                                     path,
                                 )
                             used_variables.add(args[0].this.lower())
-                elif macro_func_or_var.__class__ is d.MacroVar:
+                elif isinstance(macro_func_or_var, d.MacroVar):
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1922]: this one can stay

File: sqlmesh/core/model/definition.py [hunk 6]
```diff
@@ -1882,12 +1919,19 @@ def _python_env(
...
                     name = macro_func_or_var.name.lower()
                     if name in macros:
                         used_macros[name] = macros[name]
                     elif name in variables:
                         used_variables.add(name)
+                elif (
+                    isinstance(macro_func_or_var, exp.Identifier) and "@" in macro_func_or_var.this
+                ):
+                    for m in MacroStrTemplate.pattern.findall(macro_func_or_var.this):
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1931]: nice

File: sqlmesh/core/model/definition.py [hunk 7]
```diff
@@ -1882,12 +1919,19 @@ def _python_env(
...
+                        var_name = m[2] or m[1]
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1932]: what's m2 and m1? can you do for x, y so i can read what these things are

File: tests/core/test_model.py [hunk 8]
```diff
@@ -3759,3 +3809,28 @@ def test_this_model() -> None:
 
     assert model.render_pre_statements()[0].sql() == """COPY "db"."table" TO 'a'"""
     assert model.render_post_statements()[0].sql() == """COPY "db"."table" TO 'b'"""
+
+
+def test_macros_in_model_statement(sushi_context, assert_exp_eq):
+    expressions = d.parse(
+        """
+        MODEL (
+            name @{gateway}.test_model,
```
Comment (tobymao) [inline tests/core/test_model.py:3818]: maybe just make this a bit more complex,

name @{gate_way}_@{gate_way}.test_model

**Review by tobymao — APPROVED** [review]

File: sqlmesh/core/model/definition.py (hunk 5 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1922]: I'm confused. Why?

File: tests/core/test_model.py (hunk 8 above)
Comment (izeigerman) [reply tests/core/test_model.py:3818]: There's a combo scenario in the other test. Or did you mean to test something else?

File: sqlmesh/core/model/definition.py (hunk 5 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1922]: Just saw your other comment

File: sqlmesh/core/model/definition.py (hunk 7 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py]: unpacked the tuple

File: tests/core/test_model.py (hunk 8 above)
Comment (izeigerman) [reply tests/core/test_model.py]: Updated

**Pull request merged**